js_modules/dagit/packages/core/src/asset-graph/useLiveDataForAssetKeys.tsx

```diff
@@ -27,13 +34,126 @@ export function useLiveDataForAssetKeys(assetKeys: AssetKeyInput[]) {
     return liveResult.data ? buildLiveData(liveResult.data) : {};
   }, [liveResult.data]);
 
+  // Track whether the data is being refetched so incoming asset events don't trigger
+  // duplicate requests for live data.
+  const fetching = React.useRef(false);
+  fetching.current = [NetworkStatus.refetch, NetworkStatus.loading].includes(
+    liveResult.networkStatus,
+  );
+
+  const timerRef = React.useRef<NodeJS.Timeout | null>(null);
+
+  const onRefreshDebounced = React.useCallback(() => {
+    // This is a basic `throttle`, except that if it fires and finds the live query
+    // is already refreshing it debounces again.
+    const refetch = liveResult.refetch;
+    const fire = () => {
+      if (fetching.current) {
+        timerRef.current = setTimeout(fire, SUBSCRIPTION_MAX_POLL_RATE);
+      } else {
+        timerRef.current = null;
+        refetch();
+      }
+    };
+    if (!timerRef.current) {
+      timerRef.current = setTimeout(fire, SUBSCRIPTION_MAX_POLL_RATE);
+    }
+    return () => {
+      if (timerRef.current) {
+        clearTimeout(timerRef.current);
+      }
+    };
+  }, [timerRef, liveResult.refetch]);
+
+  // If the event log storage does not support streaming us asset events, fall back to
+  // a polling approach and trigger a single refresh when a run is launched for immediate feedback
+  const liveDataRefreshState = useQueryRefreshAtInterval(liveResult, SUBSCRIPTION_IDLE_POLL_RATE);
+  useDidLaunchEvent(liveResult.refetch, 2000);
+
+  const assetKeyTokens = React.useMemo(() => new Set(assetKeys.map(tokenForAssetKey)), [assetKeys]);
+  const assetStepKeys = React.useMemo(
+    () => new Set(...(liveResult.data?.assetNodes.flatMap((n) => n.opNames) || [])),
+    [liveResult],
+  );
+
+  const runInProgressId = uniq(
+    Object.values(liveDataByNode).flatMap((p) => [...p.unstartedRunIds, ...p.inProgressRunIds]),
+  )
+    .sort()
+    .slice(0, 3);
+
+  const runWatchers = (
+    <>
+      {runInProgressId.map((runId) => (
+        <RunLogObserver
+          runId={runId}
+          key={runId}
+          assetKeyTokens={assetKeyTokens}
+          assetStepKeys={assetStepKeys}
+          callback={onRefreshDebounced}
+        />
+      ))}
+    </>
+  );
+
   return {
-    liveResult,
     liveDataByNode,
+    liveDataRefreshState,
+    runWatchers,
     assetKeys,
   };
 }
 
+const RunLogObserver: React.FC<{
+  runId: string;
+  assetKeyTokens: Set<string>;
+  assetStepKeys: Set<string>;
+  callback: () => void;
+}> = React.memo(({runId, assetKeyTokens, assetStepKeys, callback}) => {
+  // Useful for testing this component:
+  const counter = React.useRef(0);
+  React.useEffect(() => {
+    console.log(`Subscribed to ${runId}`);
+    return () => console.log(`Unsubscribed from ${runId} after ${counter.current} messages`);
+  }, [runId]);
+
+  useSubscription<AssetLiveRunLogsSubscription>(ASSET_LIVE_RUN_LOGS_SUBSCRIPTION, {
```

Comment:
Do we end up with any unwanted extra renders due to using `useSubscription`?

Reply:
I checked and we're all good! The approach here is a bit odd (creating a leaf "component" for each subscription and using React keys to make sure the lifespan of the component matches the lifespan we want on the subscription), but when the useSubscription emits an event, only this tiny leaf node re-renders and it's just a `<span />` , so things are pretty quiet with a lot of events coming in!